src/Microsoft.AspNetCore.Mvc.Core/Routing/DispatcherUrlHelper.cs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    /// <summary>
+    /// An implementation of <see cref="IUrlHelper"/> that uses <see cref="ILinkGenerator"/> to build URLs 
+    /// for ASP.NET MVC within an application.
+    /// </summary>
+    internal class DispatcherUrlHelper : UrlHelperBase
+    {
+        private readonly ILogger<DispatcherUrlHelper> _logger;
+        private readonly ILinkGenerator _linkGenerator;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DispatcherUrlHelper"/> class using the specified
+        /// <paramref name="actionContext"/>.
+        /// </summary>
+        /// <param name="actionContext">The <see cref="Mvc.ActionContext"/> for the current request.</param>
+        /// <param name="linkGenerator">The <see cref="ILinkGenerator"/> used to generate the link.</param>
+        /// <param name="logger">The <see cref="ILogger"/>.</param>
+        public DispatcherUrlHelper(
+            ActionContext actionContext,
+            ILinkGenerator linkGenerator,
+            ILogger<DispatcherUrlHelper> logger)
+            : base(actionContext)
+        {
+            if (linkGenerator == null)
+            {
+                throw new ArgumentNullException(nameof(linkGenerator));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _linkGenerator = linkGenerator;
+            _logger = logger;
+        }
+
+        /// <inheritdoc />
+        public override string Action(UrlActionContext urlActionContext)
+        {
+            if (urlActionContext == null)
+            {
+                throw new ArgumentNullException(nameof(urlActionContext));
+            }
+
+            var valuesDictionary = GetValuesDictionary(urlActionContext.Values);
+
+            if (urlActionContext.Action == null)
+            {
+                if (!valuesDictionary.ContainsKey("action") &&
```

Comment:
Are there `const` values anywhere in MVC for "area", "controller", "action"? These strings are reused everywhere.

Reply:
Yeah, I do not see any central place for these strings. Probably there should be a file with these constants in Mvc.Abstractions.